examples/librilight/matching.py

```diff
@@ -0,0 +1,832 @@
+        start = max(start, match[0] - query_length // 4)
+        end = min(end, match[1] + query_length // 4)
+
+        target = sourced_text.binary_text[start:end]
+
+        alignment = levenshtein_distance(query=query, target=target)
+        if alignment[0] < min_cost:
+            min_cost = alignment[0]
+            best_alignment = alignment[1][0]  # Takes only the first alignment for now
+            best_match = (start, end)
+
+    # best_match is global index into sourced_text
+    # best_alignment is local index into target
+    start = best_match[0] + best_alignment[0]
+    # plus 1 here because the boundary given by best_alignment is [start, end] (end included)
+    end = best_match[0] + best_alignment[1] + 1
+
+    assert sourced_text.doc[start] == sourced_text.doc[end - 1], (
+        sourced_text.doc[start],
+        sourced_text.doc[end - 1],
+    )
+    assert sourced_text.doc[query_start] == sourced_text.doc[query_end] - 1, (
+        sourced_text.doc[query_start],
+        sourced_text.doc[query_end] - 1,
+    )
+
+    segments = split_into_segments(
+        sourced_text=sourced_text,
+        query_start=query_start,
+        target_start=start,
+        alignments=best_alignment[2],
+    )
+
+    return (cut_indexes, segments)
+
+
+def process_one_batch(
+    batch_cuts: List[MonoCut], params: AttributeDict, cuts_writer: SequentialJsonlWriter
+):
+    """
+    Process the text matching algorithm for a batch of cuts, each cuts could
+    have multiple supervisions, and write the new cuts (with some extra attributes,
+    like text, begin pos, end pos).
+
+    Args:
+      batch_cuts:
+        A list of cuts, the same as cut generated by `lhotse prepare`, except that it contains
+        recognition results (in the format of AlignItem) and the reference book path.
+        TODO: It is hard to describe the content of a cut here, we could demonstrate this in some
+        higher level documents.
+      params:
+        Configurable arguments needed during the matching.
+      cuts_writer:
+        The writes to write out new cuts.
+    """
+    # List of transcripts (total number of valid supervisions in the cuts).
+    transcripts: List[Transcript] = []
+
+    # Contains cut index and local supervision index
+    transcripts_cut_index: List[Tuple[int, int]] = []
+
+    # Constructed from the valid books in the cuts
+    books: List[TextSource] = []
+
+    # unique books exist in this batch of cuts
+    book_paths: Set[str] = set()
+    query_len = 0
+
+    # Construct transcripts
+    for i, cut in enumerate(batch_cuts):
+        # No text book available, skip this cut.
+        if cut.text_path == "":
+            logging.warning(f"Skipping {cut.id} due to missing of reference book")
+            continue
+        for j, sup in enumerate(cut.supervisions):
+            # Transcript requires the input to be the dict like this.
+            aligns = {"text": [], "begin_times": []}
```

Comment:
I suggest to use
```python3
text_list = []
begin_times_list = []
for ali in sup.alignment["symbol"]:
  text_list.append(ali.symbol)
  begin_times_list.append(ali.start)
```

dictionary look-up takes time and using a list can avoid that.